cpp/src/arrow/filesystem/path_util.cc

```diff
@@ -287,6 +288,42 @@ bool IsLikelyUri(util::string_view v) {
   return ::arrow::internal::IsValidUriScheme(v.substr(0, pos));
 }
 
+struct Globber::Impl {
+  std::regex pattern_;
+
+  explicit Impl(const std::string& p) : pattern_(std::regex(PatternToRegex(p))) {}
+
+  std::string PatternToRegex(const std::string& p) {
```

Comment:
Nit: can make this a static method
```suggestion
  static std::string PatternToRegex(const std::string& p) {
```